cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,176 @@
+      if (0 == threadIdx.x % WARP_SIZE) {
+        output.column(j).set_mask_word(word_index(i), result);
+      }
+    }
+
+    // Update active threads before branching
+    i += stride_x;
+    active_threads = __ballot_sync(active_threads, i < input.num_columns());
+  }
+}
+
+struct launch_kernel {
+  template <typename T>
+  void operator()(
+    table_view const& input,
+    mutable_table_view& output,
+    bool has_null,
+    cudaStream_t stream)
+  {
+    // Copy input columns `data` and `valid` pointers to device
+    auto device_input = table_device_view::create(input, stream);
+    auto device_output = mutable_table_device_view::create(output, stream);
+
+    dim3 dimBlock(WARP_SIZE, WARP_SIZE);
+    dim3 dimGrid(std::min(util::div_rounding_up_safe(input.num_columns(), WARP_SIZE), MAX_GRID_SIZE),
+                 std::min(util::div_rounding_up_safe(input.num_rows(), WARP_SIZE), MAX_GRID_SIZE));
+
+    gpu_transpose<T><<<dimGrid, dimBlock, 0, stream>>>(*device_input, *device_output);
+
+    if (has_null) {
+      gpu_transpose_valids<<<dimGrid, dimBlock, 0, stream>>>(*device_input, *device_output);
+
+      // Force null counts to be recomputed next time they are queried
+      for (auto& column : output) {
+        column.set_null_count(UNKNOWN_NULL_COUNT);
+      }
+    }
+  }
+};
+
+}  // namespace
+
+namespace detail {
+
+std::unique_ptr<experimental::table> transpose(table_view const& input,
+  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+{
+  auto const input_ncols = input.num_columns();
+  auto const input_nrows = input.num_rows();
+
+  // If there are no rows in the input, return successfully
+  if (input_ncols == 0 || input_nrows == 0) {
+    return std::make_unique<experimental::table>(std::vector<std::unique_ptr<column>>{});
+  }
+
+  // Check datatype homogeneity
+  auto const& first = input.column(0);
+  auto const dtype = first.type();
+  CUDF_EXPECTS(std::all_of(input.begin(), input.end(), [dtype](auto const& col) {
+    return dtype == col.type(); }), "Column type mismatch");
+
+  // TODO does this need to support non-fixed-width tables?
+  CUDF_EXPECTS(is_fixed_width(dtype), "Invalid, non-fixed-width type.");
+
+  nvtx::range_push("CUDF_TRANSPOSE", nvtx::color::GREEN);
+
+  // Check if there are nulls to be processed
+  bool const has_null = has_nulls(input);
+  auto const allocation_policy = has_null ? experimental::mask_allocation_policy::ALWAYS
+    : experimental::mask_allocation_policy::NEVER;
+
+  auto const& output_ncols = input_nrows;
+  auto const& output_nrows = input_ncols;
+
+  // Allocate output table with transposed shape
+  std::vector<std::unique_ptr<column>> out_columns(output_ncols);
+  std::generate(out_columns.begin(), out_columns.end(), [=]() {
+    return experimental::detail::allocate_like(first, output_nrows,
+      allocation_policy, mr, stream);
+  });
+  auto output = std::make_unique<experimental::table>(std::move(out_columns));
+  auto output_view = output->mutable_view();
```

Comment:
FYI, this is the point at which the null counts will be reset to `UNKNOWN_NULL_COUNT` for all columns in `output`.

Reply:
Ok, I'm going to remove the mutable view

Reply:
Never mind, actually I can't avoid creating the mutable view, as I need it in order to create the mutable_device_view.